File: hermes-plugin/pumpup_hermes/tools.py
"""The pumpup_* tool handlers, task resolution, and the dispatch wrapper that routes every error to
a structured failure (the single choke point). Handlers are async, return a JSON string, and read
the Hermes session_id from **kwargs to bind a durable Pump Up task. All share the one global client.
"""

from __future__ import annotations

import json
import threading
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Awaitable, Callable, Optional

from . import schemas
from .attachments import upload_attachments
from .client import get_client
from .config import is_configured
from .errors import to_tool_failure
from .state import Breadcrumb, get_store

from pydantic import TypeAdapter
from pumpup import ApprovalRecommendation, AsyncPumpUp, Field, FieldBid, MetadataPatchDto

TOOLSET = "pumpup"

# Build the agent's wire dicts into the SDK's typed request models, validating at construction — a malformed
# recommendation / field / bid fails locally with a precise error before the request. A TypeAdapter handles
# the Field discriminated union (each arm constructed by its `type`), which a single model class can't.
_RECOMMENDATION = TypeAdapter(Optional[ApprovalRecommendation])
_FIELD_BIDS = TypeAdapter(Optional[list[FieldBid]])
_FIELDS = TypeAdapter(list[Field])
_METADATA_PATCH = TypeAdapter(Optional[MetadataPatchDto])

# The agent guide — embedded markdown (generated artifact) returned by pumpup_guide, plus the SKILL.md
# backing the pull-loadable `pumpup:guide` skill.
_GUIDE_TEXT = (Path(__file__).parent / "generated" / "agent_guide.md").read_text(encoding="utf-8")
_GUIDE_SKILL = Path(__file__).parent / "skills" / "guide" / "SKILL.md"

# Per-process cache of the auto-created task id per session_id — keyed exactly like the idempotency key
# (one task per session), so a cache miss after a restart re-resolves the same task; the cache only spares
# a round-trip. Guarded by a lock: tool calls run on multiple worker threads.
_session_tasks: dict[str, str] = {}
_session_tasks_lock = threading.Lock()


async def resolve_task_id(client: AsyncPumpUp, project_name: str, session_id: Optional[str]) -> str:
    """Find-or-create this session's one durable task, cached in-process. Keyed by session_id so a new
    session opens a fresh task — a decision must land in the session that opened it (project of first use)."""
    if not session_id:
        raise RuntimeError("pumpup: no Hermes session_id to bind a task to")
    with _session_tasks_lock:
        cached = _session_tasks.get(session_id)
    if cached:
        return cached
    task = await client.tasks.create(
        idempotency_key=f"hermes-task:{session_id}",
        project_name=project_name,
        name=f"Hermes session {session_id}",
        external_id=session_id,
    )
    with _session_tasks_lock:
        _session_tasks[session_id] = task.id
    return task.id


async def fetch_result(client: AsyncPumpUp, kind: str, request_id: str) -> Any:
    """Fetch a request's decision: the result once decided, None while pending (the SDK maps the 204
    pending response to a None body). An unknown id / server error raises and propagates."""
    if kind == "approval":
        return await client.approvals.get_result(request_id)
    return await client.elicitations.get_result(request_id)


async def _record_action(args: dict, session_id: Optional[str]) -> str:
    """Record autonomous agent activity on the task timeline; may attach files, patch metadata, transition."""
    client = get_client()
    project_name = args["projectName"]
    task_id = await resolve_task_id(client, project_name, session_id)
    add_attachments = await upload_attachments(client, args.get("attachments"))
    created = await client.events.agent_action(
        idempotency_key=str(uuid.uuid4()),
        project_name=project_name,
        description=args["description"],
        task_id=task_id,
        detail=args.get("detail"),
        transitions_to=args.get("transitionsTo"),
        metadata_patch=_METADATA_PATCH.validate_python(args.get("metadataPatch")),
        add_attachments=add_attachments or None,
        external_trace_id=args.get("externalTraceId"),
    )
    return json.dumps({"event_id": created.event_id})


async def _report_exception(args: dict, session_id: Optional[str]) -> str:
    """Report an agent error/exception; transition-only (no metadata/attachments — those ride an action)."""
    client = get_client()
    project_name = args["projectName"]
    task_id = await resolve_task_id(client, project_name, session_id)
    created = await client.events.exception(
        idempotency_key=str(uuid.uuid4()),
        project_name=project_name,
        message=args["message"],
        task_id=task_id,
        detail=args.get("detail"),
        transitions_to=args.get("transitionsTo"),
        external_trace_id=args.get("externalTraceId"),
    )
    return json.dumps({"event_id": created.event_id})


async def _add_note(args: dict, session_id: Optional[str]) -> str:
    """Add a freeform annotation to the task timeline (no attachments / state changes)."""
    client = get_client()
    project_name = args["projectName"]
    task_id = await resolve_task_id(client, project_name, session_id)
    created = await client.events.note(
        idempotency_key=str(uuid.uuid4()),
        project_name=project_name,
        message=args["message"],
        task_id=task_id,
        external_trace_id=args.get("externalTraceId"),
    )
    return json.dumps({"event_id": created.event_id})


async def _get_decision(args: dict, _session_id: Optional[str]) -> str:
    """One-shot, non-blocking read of a request's decision; 200 = decided payload, else still pending."""
    client = get_client()
    result = await fetch_result(client, args["type"], args["requestId"])
    if result is None:
        return json.dumps({"status": "pending"})
    return json.dumps({"status": "decided", "decision": result.model_dump(mode="json", by_alias=True)})


async def _guide(args: dict, _session_id: Optional[str]) -> str:
    """Return the Pump Up usage guide (how/when to use the tools, the pending/resume model). Static text."""
    return json.dumps({"guide": _GUIDE_TEXT})


async def _attach_to_task(
    client: AsyncPumpUp, project_name: str, task_id: str, inputs: Optional[list[dict[str, Any]]]
) -> list[str]:
    """Upload attachments, attach them to the task via an action event, and return their upload ids.
    The API attaches files only through an action; a request then renders an already-attached file by id."""
    refs = await upload_attachments(client, inputs)
    if not refs:
        return []
    await client.events.agent_action(
        idempotency_key=str(uuid.uuid4()),
        project_name=project_name,
        task_id=task_id,
        description=f"Attached {len(refs)} file(s) for review",
        add_attachments=refs,
    )
    return [ref.upload_id for ref in refs]


def _park(request_id: str, kind: str, summary: str, session_id: str) -> None:
    """Durably record a request→session breadcrumb so the poll loop can resume the origin session on a decision."""
    store = get_store()
    crumb = Breadcrumb(
        request_id=request_id,
        session_id=session_id,
        type=kind,
        summary=summary,
        created_at=datetime.now(timezone.utc).isoformat(),
    )
    store.record(crumb)


async def _request_approval(args: dict, session_id: Optional[str]) -> str:
    """Request human sign-off; create the approval, durably park the wait, return pending immediately."""
    client = get_client()
    project_name = args["projectName"]
    task_id = await resolve_task_id(client, project_name, session_id)
    attachment_ids = await _attach_to_task(client, project_name, task_id, args.get("attachments"))
    summary = args["summary"]
    created = await client.approvals.create(
        idempotency_key=str(uuid.uuid4()),
        project_name=project_name,
        summary=summary,
        task_id=task_id,
        key_value_context=args.get("keyValueContext"),
        recommendation=_RECOMMENDATION.validate_python(args.get("recommendation")),
        attachments=attachment_ids or None,
    )
    _park(created.event_id, "approval", summary, session_id)
    return json.dumps({"requestId": created.event_id, "status": "pending"})


async def _request_elicitation(args: dict, session_id: Optional[str]) -> str:
    """Ask a human for structured fields; create the elicitation, durably park the wait, return pending."""
    client = get_client()
    project_name = args["projectName"]
    task_id = await resolve_task_id(client, project_name, session_id)
    attachment_ids = await _attach_to_task(client, project_name, task_id, args.get("attachments"))
    summary = args["summary"]
    created = await client.elicitations.create(
        idempotency_key=str(uuid.uuid4()),
        project_name=project_name,
        summary=summary,
        task_id=task_id,
        fields=_FIELDS.validate_python(args["fields"]),
        key_value_context=args.get("keyValueContext"),
        recommendation=_FIELD_BIDS.validate_python(args.get("recommendation")),
        attachments=attachment_ids or None,
    )
    _park(created.event_id, "elicitation", summary, session_id)
    return json.dumps({"requestId": created.event_id, "status": "pending"})


def _tool_handler(action: str, core: Callable[[dict, Optional[str]], Awaitable[str]]) -> Callable[..., Awaitable[str]]:
    """Wrap a tool's core coroutine: pass the dispatched session_id, map any error to a failure JSON."""

    async def handler(args: dict, **kwargs: Any) -> str:
        try:
            return await core(args, kwargs.get("session_id"))
        except Exception as exc:  # noqa: BLE001 — the choke point; nothing should escape into Hermes
            return to_tool_failure(action, exc)

    return handler


def _register(ctx: Any, schema: dict, handler: Callable[..., Awaitable[str]]) -> None:
    """Register one async tool into the `pumpup` toolset, gated on Pump Up being configured."""
    ctx.register_tool(
        name=schema["name"],
        toolset=TOOLSET,
        schema=schema,
        handler=handler,
        check_fn=is_configured,
        is_async=True,
        description=schema["description"],
    )


def register_request_tools(ctx: Any) -> None:
    """Register the two request tools (approval, elicitation): create → park breadcrumb → return pending."""
    _register(ctx, schemas.REQUEST_APPROVAL, _tool_handler("request_approval", _request_approval))
    _register(ctx, schemas.REQUEST_ELICITATION, _tool_handler("request_elicitation", _request_elicitation))


def register_capture_tools(ctx: Any) -> None:
    """Register the fire-and-forget timeline tools plus the one-shot decision read (no wait, no resume)."""
    _register(ctx, schemas.RECORD_ACTION, _tool_handler("record_action", _record_action))
    _register(ctx, schemas.REPORT_EXCEPTION, _tool_handler("report_exception", _report_exception))
    _register(ctx, schemas.ADD_NOTE, _tool_handler("add_note", _add_note))
    _register(ctx, schemas.GET_DECISION, _tool_handler("get_decision", _get_decision))


def register_guide(ctx: Any) -> None:
    """Register the pumpup_guide tool and, when configured, the pull-loadable `pumpup:guide` skill."""
    _register(ctx, schemas.GUIDE, _tool_handler("guide", _guide))
    if is_configured():
        ctx.register_skill("guide", _GUIDE_SKILL, description=schemas.GUIDE["description"])

File: openclaw-plugin/src/tools.ts
import { randomUUID } from "node:crypto";

import { jsonResult } from "openclaw/plugin-sdk/core";
import type { AnyAgentTool, OpenClawPluginApi, OpenClawPluginToolContext } from "openclaw/plugin-sdk/plugin-entry";
import { type Static, type TSchema, Type } from "typebox";

import type { PumpUp, PumpUpClient } from "@pumpupai/pumpup-sdk";
import { type AttachmentInput, uploadAttachments } from "./attachments.js";
import type { PumpUpPluginConfig } from "./config.js";
import { toToolFailure } from "./errors.js";
import { agentGuide } from "./generated/agent-guide.js";
import {
  approvalRecommendationSchema,
  fieldBidSchema,
  fieldSchema,
  metadataPatchSchema,
} from "./generated/contract-schemas.js";
import { buildFlowState, type PumpUpRequestType } from "./state.js";

const AttachmentParam = Type.Object({
  path: Type.String({ description: "Workspace-relative file path" }),
  label: Type.String({ description: "Display label shown to the human reviewer" }),
});

const KeyValueContext = Type.Optional(
  Type.Record(Type.String(), Type.String(), { description: "Context shown to the reviewer, as key→value" }),
);

const DetailContext = Type.Optional(
  Type.Record(Type.String(), Type.Unknown(), { description: "Freeform structured detail recorded on the event" }),
);

/// SDK contract schemas (generated from openapi.yaml → SDK), typed back to their SDK types. The
/// agent-declared elicitation field union is the canonical `Field` schema; recommendations mirror
/// the SDK's `ApprovalRecommendation` / `FieldBid`; `MetadataPatch` mirrors `MetadataPatchDto`.
const ElicitationField = Type.Unsafe<PumpUp.Field>(fieldSchema);
const ApprovalRecommendationSchema = Type.Unsafe<PumpUp.ApprovalRecommendation>(approvalRecommendationSchema);
const FieldBidSchema = Type.Unsafe<PumpUp.FieldBid>(fieldBidSchema);
const MetadataPatch = Type.Unsafe<PumpUp.MetadataPatchDto>(metadataPatchSchema);

const ApprovalParams = Type.Object({
  projectName: Type.String({ description: "Pump Up project slug" }),
  summary: Type.String({ description: "Human-readable description of what needs approval" }),
  keyValueContext: KeyValueContext,
  recommendation: Type.Optional(ApprovalRecommendationSchema),
  attachments: Type.Optional(Type.Array(AttachmentParam)),
});

const ElicitationParams = Type.Object({
  projectName: Type.String({ description: "Pump Up project slug" }),
  summary: Type.String({ description: "Human-readable description of the input needed" }),
  fields: Type.Array(ElicitationField, { description: "Fields to elicit; each is validated per type server-side" }),
  keyValueContext: KeyValueContext,
  recommendation: Type.Optional(Type.Array(FieldBidSchema)),
  attachments: Type.Optional(Type.Array(AttachmentParam)),
});

/// Fields shared by both task-event tools (action + exception). `detail` is freeform; `transitionsTo`
/// mutates the target task and must match its declared step graph.
const EventTaskFields = {
  detail: DetailContext,
  transitionsTo: Type.Optional(Type.String({ description: "Declared task state to transition into" })),
  externalTraceId: Type.Optional(Type.String({ description: "Links this event to your agent's trace/run id" })),
};

/// Governed-data fields carried only by `action` — the API drops them from `exception` (transition-only):
/// `metadataPatch` mutates the task bag; `attachments` are workspace files attached to the case.
const ActionGovernedFields = {
  metadataPatch: Type.Optional(MetadataPatch),
  attachments: Type.Optional(Type.Array(AttachmentParam)),
};

const RecordActionParams = Type.Object({
  projectName: Type.String({ description: "Pump Up project slug" }),
  description: Type.String({ description: "What the agent did" }),
  ...EventTaskFields,
  ...ActionGovernedFields,
});

const ExceptionParams = Type.Object({
  projectName: Type.String({ description: "Pump Up project slug" }),
  message: Type.String({ description: "What went wrong" }),
  ...EventTaskFields,
});

const NoteParams = Type.Object({
  projectName: Type.String({ description: "Pump Up project slug" }),
  message: Type.String({ description: "The annotation text" }),
  externalTraceId: Type.Optional(Type.String({ description: "Links this event to your agent's trace/run id" })),
});

const GetDecisionParams = Type.Object({
  requestId: Type.String({ description: "The requestId returned by request_approval / request_elicitation" }),
  type: Type.Union([Type.Literal("approval"), Type.Literal("elicitation")], {
    description: "Which kind of request this id refers to",
  }),
});

/// Register the request tools that POST to Pump Up and durably park the wait as a TaskFlow.
export function registerRequestTools(api: OpenClawPluginApi, cfg: PumpUpPluginConfig, client: PumpUpClient): void {
  api.registerTool((ctx) => buildApprovalTool(api, cfg, client, ctx));
  api.registerTool((ctx) => buildElicitationTool(api, cfg, client, ctx));
}

/// Register the fire-and-forget timeline tools plus the one-shot decision read (no flow, no wait).
export function registerCaptureTools(api: OpenClawPluginApi, client: PumpUpClient): void {
  api.registerTool((ctx) => buildRecordActionTool(client, ctx));
  api.registerTool((ctx) => buildReportExceptionTool(client, ctx));
  api.registerTool((ctx) => buildAddNoteTool(client, ctx));
  api.registerTool(() => buildGetDecisionTool(client));
}

/// Register the read-only guide tool — the agent's pull path to the holistic Pump Up usage model.
export function registerGuideTool(api: OpenClawPluginApi): void {
  api.registerTool(() => buildGuideTool());
}

/// Return the Pump Up usage guide (how/when to use the tools, the pending/resume model). Static, no I/O.
function buildGuideTool(): AnyAgentTool {
  return defineTool({
    name: "pumpup_guide",
    label: "Pump Up guide",
    description: "How and when to use the Pump Up tools — read this before your first approval/elicitation.",
    parameters: Type.Object({}),
    execute: async () => jsonResult({ guide: agentGuide }),
  });
}

/// Approval request tool: POST → park flow as waiting → return the pending handle.
function buildApprovalTool(
  api: OpenClawPluginApi,
  cfg: PumpUpPluginConfig,
  client: PumpUpClient,
  ctx: OpenClawPluginToolContext,
): AnyAgentTool {
  return defineTool({
    name: "pumpup_request_approval",
    label: "Request approval",
    description:
      "Request human sign-off in Pump Up. Returns pending, then end your turn — the decision later resumes "
      + "this session (don't poll). New to Pump Up? Call pumpup_guide first.",
    parameters: ApprovalParams,
    execute: async (params) => {
      const taskId = await resolveTaskId(client, params.projectName, ctx);
      const attachments = await attachToTask(client, ctx, params.projectName, taskId, params.attachments);
      const created = await client.approvals.create({
        "Idempotency-Key": randomUUID(),
        projectName: params.projectName,
        summary: params.summary,
        taskId,
        keyValueContext: params.keyValueContext,
        recommendation: params.recommendation,
        attachments: attachments.length > 0 ? attachments : undefined,
      });
      const flowId = parkRequest(api, cfg, "approval", created.eventId, params.summary, ctx);
      return jsonResult({ requestId: created.eventId, flowId, status: "pending" });
    },
  });
}

/// Elicitation request tool: POST agent-declared fields → park flow as waiting → return pending handle.
function buildElicitationTool(
  api: OpenClawPluginApi,
  cfg: PumpUpPluginConfig,
  client: PumpUpClient,
  ctx: OpenClawPluginToolContext,
): AnyAgentTool {
  return defineTool({
    name: "pumpup_request_elicitation",
    label: "Request elicitation",
    description:
      "Ask a human for structured input in Pump Up. Returns pending, then end your turn — the answer later "
      + "resumes this session (don't poll). New to Pump Up? Call pumpup_guide first.",
    parameters: ElicitationParams,
    execute: async (params) => {
      const taskId = await resolveTaskId(client, params.projectName, ctx);
      const attachments = await attachToTask(client, ctx, params.projectName, taskId, params.attachments);
      const created = await client.elicitations.create({
        "Idempotency-Key": randomUUID(),
        projectName: params.projectName,
        summary: params.summary,
        fields: params.fields,
        taskId,
        keyValueContext: params.keyValueContext,
        recommendation: params.recommendation,
        attachments: attachments.length > 0 ? attachments : undefined,
      });
      const flowId = parkRequest(api, cfg, "elicitation", created.eventId, params.summary, ctx);
      return jsonResult({ requestId: created.eventId, flowId, status: "pending" });
    },
  });
}

/// Record autonomous agent activity on the task timeline; may attach files, patch metadata, transition state.
function buildRecordActionTool(client: PumpUpClient, ctx: OpenClawPluginToolContext): AnyAgentTool {
  return defineTool({
    name: "pumpup_record_action",
    label: "Record action",
    description: "Record something the agent did on the Pump Up timeline. Fire-and-forget; returns the event id.",
    parameters: RecordActionParams,
    execute: async (params) => {
      const taskId = await resolveTaskId(client, params.projectName, ctx);
      const addAttachments = await uploadAttachments(client, ctx.workspaceDir, params.attachments);
      const created = await client.events.agentAction({
        "Idempotency-Key": randomUUID(),
        projectName: params.projectName,
        description: params.description,
        taskId,
        detail: params.detail,
        metadataPatch: params.metadataPatch,
        transitionsTo: params.transitionsTo,
        addAttachments: addAttachments.length > 0 ? addAttachments : undefined,
        externalTraceId: params.externalTraceId,
      });
      return jsonResult({ eventId: created.eventId });
    },
  });
}

/// Report something that went wrong; transition-only (no metadata/attachments — those ride `action`).
function buildReportExceptionTool(client: PumpUpClient, ctx: OpenClawPluginToolContext): AnyAgentTool {
  return defineTool({
    name: "pumpup_report_exception",
    label: "Report exception",
    description: "Report an agent error/exception on the Pump Up timeline. Fire-and-forget; returns the event id.",
    parameters: ExceptionParams,
    execute: async (params) => {
      const taskId = await resolveTaskId(client, params.projectName, ctx);
      const created = await client.events.exception({
        "Idempotency-Key": randomUUID(),
        projectName: params.projectName,
        message: params.message,
        taskId,
        detail: params.detail,
        transitionsTo: params.transitionsTo,
        externalTraceId: params.externalTraceId,
      });
      return jsonResult({ eventId: created.eventId });
    },
  });
}

/// Add a freeform annotation to the task timeline (no attachments / state changes).
function buildAddNoteTool(client: PumpUpClient, ctx: OpenClawPluginToolContext): AnyAgentTool {
  return defineTool({
    name: "pumpup_add_note",
    label: "Add note",
    description: "Add a freeform note to the Pump Up timeline. Fire-and-forget; returns the event id.",
    parameters: NoteParams,
    execute: async (params) => {
      const taskId = await resolveTaskId(client, params.projectName, ctx);
      const created = await client.events.note({
        "Idempotency-Key": randomUUID(),
        projectName: params.projectName,
        message: params.message,
        taskId,
        externalTraceId: params.externalTraceId,
      });
      return jsonResult({ eventId: created.eventId });
    },
  });
}

/// One-shot, non-blocking read of a request's decision; 200 = decided payload, 204 = still pending.
function buildGetDecisionTool(client: PumpUpClient): AnyAgentTool {
  return defineTool({
    name: "pumpup_get_decision",
    label: "Get decision",
    description: "Check once whether a Pump Up request has been decided yet. Non-blocking; does not resume.",
    parameters: GetDecisionParams,
    execute: async (params) => {
      const response = params.type === "approval"
        ? await client.approvals.getResult({ id: params.requestId }).withRawResponse()
        : await client.elicitations.getResult({ id: params.requestId }).withRawResponse();
      return response.rawResponse.status === 200
        ? jsonResult({ status: "decided", decision: response.data })
        : jsonResult({ status: "pending" });
    },
  });
}

/// Per-process cache of the auto-created task id per `sessionId` — keyed exactly like the idempotency key
/// (one task per session), so a cache miss after a restart re-resolves to the same task; the cache only
/// spares the repeat round-trip, it isn't the source of truth.
const sessionTasks = new Map<string, string>();

/// Find-or-create this session's one durable task, cached in-process. Keyed by `sessionId` so a roll/`/reset`
/// starts a fresh task — correct while rehydration is deferred (a decision must land in the session that opened
/// it, in the project of first use). The plugin owns the task lifecycle so the agent never handles task ids.
async function resolveTaskId(client: PumpUpClient, projectName: string, ctx: OpenClawPluginToolContext): Promise<string> {
  if (!ctx.sessionId) {
    throw new Error("pumpup: no session to create a task for");
  }
  const cached = sessionTasks.get(ctx.sessionId);
  if (cached) {
    return cached;
  }
  const task = await client.tasks.create({
    "Idempotency-Key": `openclaw-task:${ctx.sessionId}`,
    projectName,
    name: `OpenClaw session ${ctx.sessionKey ?? ctx.sessionId}`,
    externalId: ctx.sessionId,
  });
  sessionTasks.set(ctx.sessionId, task.id);
  return task.id;
}

/// Upload workspace files, attach them to the task via an `action` event, then return their upload ids
/// — the API attaches files only through `action`, and a request renders an already-attached file by id.
async function attachToTask(
  client: PumpUpClient,
  ctx: OpenClawPluginToolContext,
  projectName: string,
  taskId: string,
  inputs: AttachmentInput[] | undefined,
): Promise<string[]> {
  const refs = await uploadAttachments(client, ctx.workspaceDir, inputs);
  if (refs.length === 0) {
    return [];
  }
  await client.events.agentAction({
    "Idempotency-Key": randomUUID(),
    projectName,
    taskId,
    description: `Attached ${refs.length} file(s) for review`,
    addAttachments: refs,
  });
  return refs.map((ref) => ref.uploadId);
}

/// Create a managed TaskFlow for the request and transition it to `waiting` (durable until decided).
function parkRequest(
  api: OpenClawPluginApi,
  cfg: PumpUpPluginConfig,
  type: PumpUpRequestType,
  eventId: string,
  summary: string,
  ctx: OpenClawPluginToolContext,
): string {
  const flow = api.runtime.tasks.managedFlows.bindSession({ sessionKey: cfg.ownerSessionKey });
  const state = buildFlowState(eventId, type, { sessionId: ctx.sessionId, sessionKey: ctx.sessionKey, agentId: ctx.agentId });
  const created = flow.tryCreateManaged({ controllerId: "pumpup", goal: `${type}: ${summary}`, stateJson: state });
  if (!created) {
    throw new Error("pumpup: could not create a TaskFlow for the request");
  }
  const waited = flow.setWaiting({ flowId: created.flowId, expectedRevision: created.revision });
  if (!waited.applied) {
    throw new Error(`pumpup: could not park the request flow as waiting (${waited.code})`);
  }
  return created.flowId;
}

/// Build an agent tool from a TypeBox-typed spec, erasing param types at the registry boundary and
/// mapping any thrown error into a structured tool failure (single choke point for all tools).
function defineTool<S extends TSchema>(spec: {
  name: string;
  label: string;
  description: string;
  parameters: S;
  execute: (params: Static<S>) => Promise<ReturnType<typeof jsonResult>>;
}): AnyAgentTool {
  return {
    name: spec.name,
    label: spec.label,
    description: spec.description,
    parameters: spec.parameters,
    execute: async (_toolCallId: string, params: unknown) => {
      try {
        return await spec.execute(params as Static<S>);
      } catch (error) {
        return toToolFailure(spec.name, error);
      }
    },
  } as AnyAgentTool;
}
